components/storage/blockdevice/COMPONENT_SPIF/SPIFBlockDevice.cpp

```diff
@@ -540,8 +540,9 @@ spif_bd_error SPIFBlockDevice::_spi_send_program_command(int prog_inst, const vo
 
 spif_bd_error SPIFBlockDevice::_spi_send_erase_command(int erase_inst, bd_addr_t addr, bd_size_t size)
 {
-    addr = (((int)addr) & 0x00FFF000);
-    _spi_send_general_command(erase_inst, addr, NULL, 0, NULL, 0);
+    bd_addr_t BLOCK_MASK = 0x00FFF000;
+    bd_addr_t erase_addr = (addr & BLOCK_MASK);
+    _spi_send_general_command(erase_inst, erase_addr, NULL, 0, NULL, 0);
```

Comment:
I really don't see how this changes behaviour, at least on an ARM, where `int` is 32-bit. It would if the mask was 0xFFFFFF00, where you might get sign extension into the top half, but with that mask value, I don't see how you get a different result.

Presumably the code should really be 

    bd_addr_t erase_addr = addr & 0xFFFFFFFFFFFFF000;

or

    bd_addr_t erase_addr = addr & ~bdaddr_t(0xFFF);

As you don't deliberately want/need to clear top bits here, do you?

I see this is being touched in #10779, which is extending the mask to 0xFFFFFF00.

Shouldn't there be some sort of range check somewhere? Nothing in existing code or #10779 catches overflow of 3 or 4 byte address. This is a pretty public-facing API, and that could lead to silent storage corruption on too-big media.

Reply:
> Presumably the code should really be
> 
> `bd_addr_t erase_addr = addr & 0xFFFFFFFFFFFFF000;`

Agreed, totally forgot about the top bits. I'll make the change when I get in tomorrow.

My toolchain I'm using is GCC_ARM 7.3.1, so agreed perhaps `int` is indeed fine as it is and likely 32. However I do know that the address after masking my argument was definitely changed to just 0. I'll try to look deeper into this.

But still, going back to #10779 I think it makes sense to extend this flag for the top bits of bigger flash addresses.